src/hybrid_crypto/kem.rs

```diff
@@ -1,84 +1,42 @@
-use crate::{
-    asymmetric::{ristretto::X25519Crypto, AsymmetricCrypto, KeyPair},
-    hybrid_crypto::{Error, Kem},
-    kdf, KeyTrait,
-};
+use crate::{asymmetric::KeyPair, Error, KeyTrait};
 use rand_core::{CryptoRng, RngCore};
-use std::convert::TryFrom;
 
-const HKDF_INFO: &[u8; 21] = b"ecies-ristretto-25519";
-
-impl Kem for X25519Crypto {
-    type KeyPair = <Self as AsymmetricCrypto>::KeyPair;
-
-    const SECRET_KEY_LENGTH: usize = 256;
+/// Key Encapsulation Method (KEM). It is used to generate a secret key along
+/// with its encapsulation. This secret key is then typically used in a DEM
+/// scheme.
+///
+/// TODO: should the KDF used be specified here?
+pub trait Kem {
+    /// Asymmetric key pair
+    type KeyPair: KeyPair;
 
+    /// Number of bytes of the encapsulation
     const ENCAPSULATION_SIZE: usize = <Self::KeyPair as KeyPair>::PublicKey::LENGTH;
 
-    fn description() -> String {
-        todo!()
-    }
+    /// Describe the scheme in plaintext
+    fn description() -> String;
 
-    fn key_gen<R: RngCore + CryptoRng>(rng: &mut R) -> Self::KeyPair {
-        Self::KeyPair::new(rng)
-    }
+    /// Generate an asymmetric key pair
+    fn key_gen<R: RngCore + CryptoRng>(rng: &mut R) -> Self::KeyPair;
 
+    /// Return `(K, E)` the secret key and its encapsulation.
+    ///
+    /// - `pk`  : public key
+    /// - `secret_key_length`: the size in bytes of the generated secret key
```

Comment:
This has 2 effects:
 -the same KEM cannot be used for different key sizes (that is OK)
- const generics must be deployed on all objects used as dependencies such as keys (because cont generics can only be initialized from another const generics or a hard-coded value). I looked at it, quite a lot of work

Reply:
ok